snitun/metrics/collector.py

```diff
@@ -0,0 +1,130 @@
+"""Default in-memory metrics collector implementation."""
+
+from datetime import UTC, datetime
+import os
+import socket
+from typing import Any
+
+from .base import MetricsCollector
+
+
+class DefaultMetricsCollector(MetricsCollector):
+    """
+    Default in-memory metrics collector with automatic system tagging.
+
+    This collector stores metrics in memory for debugging and testing.
+    It automatically adds hostname and process_id tags to all metrics.
+    Does not need to be thread-safe as it runs in a single event loop.
+    """
+
+    def __init__(
+        self,
+        hostname: str | None = None,
+        process_id: int | None = None,
+    ) -> None:
+        """
+        Initialize the collector.
+
+        Args:
+            hostname: Override system hostname (default: socket.gethostname())
+            process_id: Override process ID (default: os.getpid())
+        """
+        self._hostname = hostname or socket.gethostname()
+        self._process_id = process_id or os.getpid()
+        self._metrics: dict[str, Any] = {}
+        self._start_time = datetime.now(tz=UTC)
+
+    def _add_system_tags(self, tags: dict[str, str] | None) -> dict[str, str]:
+        """Add hostname and process_id to all metrics."""
+        final_tags = tags.copy() if tags else {}
+        # System tags override any user-provided tags with the same name
+        final_tags["hostname"] = self._hostname
+        final_tags["process_id"] = str(self._process_id)
+        return final_tags
+
+    def _make_key(self, name: str, tags: dict[str, str]) -> str:
```

Comment:
Performance would be better if tags where a tuple of KVs that could be hashed and than `@lru_cache` a separate `_make_cached_key` function that doesn't include `self`

Reply:
This is just a test collector, it won't actually be used. A client will be dropped in here for StatsD.